### PR TITLE
sql: better `SHOW RANGES` and changes to `crdb_internal.ranges{,_no_leases}`

### LINKED ISSUE #80906: sql: `SHOW RANGES` does not show start/end keys
**Describe the problem**

The `SHOW RANGES` command appears to not emit the `start_key` and `end_key` of the ranges. They are always `NULL`.

**To Reproduce**

```
root@localhost:26257/defaultdb> SHOW RANGES FROM DATABASE system;
            table_name            | start_key | end_key | range_id | range_size_mb | lease_holder |               lease_holder_locality               |  replicas   |                                                                                                                          replica_localities
----------------------------------+-----------+---------+----------+---------------+--------------+---------------------------------------------------+-------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  comments                        | NULL      | NULL    |       20 |             0 |            4 | cloud=gce,region=europe-west1,zone=europe-west1-b | {1,2,3,4,5} | {"cloud=gce,region=europe-west1,zone=europe-west1-b","cloud=gce,region=europe-west1,zone=europe-west1-b","cloud=gce,region=europe-west1,zone=europe-west1-b","cloud=gce,region=europe-west1,zone=europe-west1-b","cloud=gce,region=europe-west1,zone=europe-west1-b"}
```

This is because it strips the table prefix from the keys, but the key bounds in a single-range table are just e.g. `/Table/50` and `/Table/51`, so the entire key is stripped, yielding `NULL`.

This is really misleading. We should show the complete key. Discussed in https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1660947554266819.

**Additional data / screenshots**

Seen on both `master`, 21.2.9 and 21.2.10.

Jira issue: CRDB-15457

Epic CRDB-22701

### LINKED ISSUE #93198: multitenant: rethink UX/schema of `crdb_internal.range{,_no_leases}` under multi-tenancy
**Describe the problem**

With https://github.com/cockroachdb/cockroach/pull/92131, secondary tenants are able to access `crdb_internal.ranges{,_no_leases}`.  Here's how some sample output looks:

```
select * from crdb_internal.ranges_no_leases;

  range_id | start_key |    start_pretty     | end_key  |     end_pretty      | table_id | database_name | schema_name | table_name | index_name | replicas |    replica_localities    | voting_replicas | non_voting_replicas | learner_replicas |   split_enforced_until
-----------+-----------+---------------------+----------+---------------------+----------+---------------+-------------+------------+------------+----------+--------------------------+-----------------+---------------------+------------------+----------------------------
        55 | \xfe8a    | /Tenant/2           | \xfe8af2 | /Tenant/2/Table/106 |        0 |               |             |            |            | {1}      | {"region=us-east1,az=b"} | {1}             | {}                  | {}               | 2022-12-07 17:23:50.37541
        56 | \xfe8af2  | /Tenant/2/Table/106 | \xffff   | /Max                |      106 | movr          | public      | users      |            | {1}      | {"region=us-east1,az=b"} | {1}             | {}                  | {}               | NULL
(2 rows)
```

The schema here assumes that every range will belong to at-most 1 table. However, this isn't true for secondary tenants, as we no longer split on table boundaries for them. As such, this schema doesn't make much sense and could use a rethink. 

----

Completing the example above, the output above seems to indicate that the `users` table is on range 56. This isn't true however, as the output of `SHOW TABLES` is here:

```
 SHOW TABLES;

  schema_name |         table_name         | type  | owner | estimated_row_count | locality
--------------+----------------------------+-------+-------+---------------------+-----------
  public      | promo_codes                | table | demo  |                1000 | NULL
  public      | rides                      | table | demo  |                 500 | NULL
  public      | user_promo_codes           | table | demo  |                   5 | NULL
  public      | users                      | table | demo  |                  50 | NULL
  public      | vehicle_location_histories | table | demo  |                1000 | NULL
  public      | vehicles                   | table | demo  |                  15 | NULL
```

All these tables are on r56. 

cc @knz @ecwall 

Jira issue: CRDB-22216

Epic CRDB-22701

### LINKED ISSUE #93617: sql: SHOW RANGES broken by coalesced ranges
**Describe the problem**

The output of SHOW RANGES assumes that a range maps to at most one db, table, index.

Since recently, ranges coalesce across tables and even databases. This means that the db/table/index name fields are now incorrect; they merely describe the data at the beginning of the range but do not include data/tables that's in the middle or at the end.

This has hilarious effects:

```sql
> create table db1; create table db1.t as select 1 as x; 
> create table db2; create table db2.t as select 1 as x; 
> show ranges from database d1;
  table_name | start_key | end_key | range_id |       range_size_mb        | lease_holder | lease_holder_locality | replicas |    replica_localities
-------------+-----------+---------+----------+----------------------------+--------------+-----------------------+----------+---------------------------
  t          | /108      | NULL    |       56 | 0.000068000000000000000000 |            1 | region=us-east1,az=b  | {1}      | {"region=us-east1,az=b"}

> show ranges from database db2;

SHOW RANGES 0
```

oops!
(the problem is because the database_name field in the range is computed on the start key, and there's no separate range with `db2` as its database_name)

Additionally, the start_key/end_key values are often broken, because they don't fit within the keyspace of the selected index (in `show ranges from table` / `from index`).

**Expected behavior**

Fundamentally the assumption that "1 range <-> at most one object" is broken so we can't repair this with the current schema of the `show ranges` output, nor the underlying `crdb_internal.ranges`. Both need to change.

We have two alternatives:

- keep the guarantee "1 row per range" (range_id remains unique), but then include a _list_ of designated objects elsewhere on the row, e.g. via a new column with type ARRAY. 
- keep the guarantee "at least 1 row per object" (db/table/index names remain unique), but then introduce synthetic range rows, 1 per SQL object included in the range, with synthetic start/end keys that map to just the object selected, and the range_id repeated across rows.

Given how we're using this facility in tests, it looks like both are desirable!

So what can happen is likely that we'll split the vtable in two: one with the former behavior and one with the latter.

Jira issue: CRDB-22439

Epic CRDB-22701

### PR DESCRIPTION
**Pull request by knz**
Fixes #93617.
Fixes #80906.
Fixes #93198.
Epic: CRDB-22701

The output of `crdb_internal.ranges{,_no_leases}` and `SHOW
RANGES` was irreparably broken by the introduction of range coalescing
(ranges spanning multiple tables/databases).

Moreover, the start/end keys of SHOW RANGES were often empty or
NULL due to incorrect/excessive truncation.

This commit fixes this by introducing a new design for SHOW RANGES
and tweaking the definition of `crdb_internal.ranges{,_no_leases}`.

Note: THIS IS A BREAKING CHANGE. See the "backward-incompatible
change" release notes below for suggestions on updating client code.

----

**Short documentation.**

<details>

The revised syntax is now:
```
  SHOW CLUSTER RANGES [WITH <options>]
  SHOW RANGES [FROM DATABASE <dbname> | FROM CURRENT_CATALOG] [ WITH <options> ]
  SHOW RANGES FROM TABLE <tablename> [ WITH <options> ]
  SHOW RANGES FROM INDEX <idxname> [ WITH <options> ]

  <options> is a combination of TABLES, INDEXES, KEYS, DETAILS and EXPLAIN.
```

New syntax: `SHOW CLUSTER RANGES`, `SHOW RANGES` with no `FROM`, `FROM
CURRENT_CATALOG`, `WITH` clause.

In summary, we have:
- `SHOW CLUSTER RANGES` which includes all ranges, including those not
  belonging to any table.
- `SHOW RANGES [FROM DATABASE | FROM CURRENT_CATALOG]` which includes
  only ranges overlapping with any table in the target db.
  Note: `SHOW RANGES` without target (NEW!) is an alias for `SHOW RANGES FROM
  CURRENT_CATALOG`.
- `SHOW RANGES FROM TABLE` selects only ranges that overlap with the
  given table.
- `SHOW RANGES FROM INDEX` selects only ranges that overlap with the
  given index.

Then:

- if `WITH TABLES` is specified, the rows are duplicated to detail
  each table included in each range (1 row per range-table
  intersection).
- if `WITH INDEXES` is specified, the rows are duplicated to detail each
  index included in each range (1 row per range-index intersection).
- otherwise, there is just 1 row per range.

In summary:

| Statement                                          | Row identity                     | Before                     | After                            |
|----------------------------------------------------|----------------------------------|----------------------------|----------------------------------|
| `SHOW RANGES FROM DATABASE`                        | rangeID                          | Includes schema/table name | (CHANGE) No schema/table name    |
| `SHOW RANGES FROM TABLE`                           | rangeID                          | Includes index name        | (CHANGE) No index name           |
| `SHOW RANGES FROM INDEX`                           | rangeID                          | Includes index name        | Unchanged                        |
| `SHOW RANGES FROM DATABASE ... WITH TABLES` (NEW)  | rangeID, schema/table name       | N/A                        | Includes schema/table name       |
| `SHOW RANGES FROM DATABASE ... WITH INDEXES` (NEW) | rangeID, schema/table/index name | N/A                        | Includes schema/table/index name |
| `SHOW RANGES FROM TABLE ... WITH INDEXES` (NEW)    | rangeID, index name              | N/A                        | Includes index name              |
| `SHOW CLUSTER RANGES` (NEW)                        | rangeID                          | N/A                        |                                  |
| `SHOW CLUSTER RANGES WITH TABLES` (NEW)            | rangeID, schema/table name       | N/A                        | Includes db/schema/table name    |
| `SHOW CLUSTER RANGES WITH INDEXES` (NEW)           | rangeID, schema/table/index name | N/A                        | Includes db/sch/table/index name |

| Statement                                          | Start/end key column, before | Start/end key column, after           |
|----------------------------------------------------|------------------------------|---------------------------------------|
| `SHOW RANGES FROM DATABASE`                        | Truncates table/index IDs    | (CHANGE) Includes table/index ID      |
| `SHOW RANGES FROM TABLE`                           | Truncates table/index IDs    | (CHANGE) Includes table/index ID      |
| `SHOW RANGES FROM INDEX`                           | Truncates table/index IDs    | Unchanged                             |
| `SHOW RANGES FROM DATABASE ... WITH TABLES` (NEW)  | N/A                          | Includes table/index ID               |
| `SHOW RANGES FROM DATABASE ... WITH INDEXES` (NEW) | N/A                          | Includes table/index ID               |
| `SHOW RANGES FROM TABLE ... WITH INDEXES` (NEW)    | N/A                          | Truncates table ID, includes index ID |
| `SHOW CLUSTER RANGES` (NEW)                        | N/A                          | Includes table/index ID               |
| `SHOW CLUSTER RANGES WITH TABLES` (NEW)            | N/A                          | Includes table/index ID               |
| `SHOW CLUSTER RANGES WITH INDEXES` (NEW)           | N/A                          | Includes table/index ID               |

In any case, all the columns from `crdb_internal.ranges_no_leases` are
included. By default, the start/end key boundaries are pretty-printed
as in previous versions.

Then:
- if `WITH KEYS` is specified, the raw key bytes are exposed alongside
  the pretty-printed key boundaries.
- if `WITH DETAILS` is specified, extra _expensive_ information is
  included in the result, as of `crdb_internal.ranges`.
  (requires more roundtrips; makes the operation slower overall)

Then:
- if `WITH EXPLAIN` is specified, the statement simply returns the
  text of the SQL query it would use if `WITH EXPLAIN` was not
  specified. This can be used for learning or troubleshooting.

</details>

See text of release notes below for more details; also the explanatory
comment at the top of `pkg/sql/delegate/show_ranges.go`.

----

**Example use.**

<details>

To test this, use for example the following setup:

```
> -- Enable merge of adjacent ranges with same zone config.
> set cluster setting spanconfig.host_coalesce_adjacent.enabled = true;
> -- Table t has two indexes with some split points.
> create table t(x int primary key, y int);
> create index sec_idx on t(y);
> alter index t@primary split at values(3);
> alter index t@sec_idx split at values(3);
> -- Tables u and v share a range with t@sec_idx.
> create table u(x int);
> create table v(x int);
> -- Make some other tables with forced split points due to different
> -- zone configs.
> create schema otherschema;
> create table otherschema.w(x int);
> create table otherschema.z(x int);
> alter table otherschema.w configure zone using num_replicas = 5;
> alter table otherschema.z configure zone using num_replicas = 7;
```

Example output for `SHOW RANGES FROM DATABASE`:

```
> show ranges from database defaultdb; -- 1 row per range
> show ranges from current_catalog; -- implicit db from session

    start_key    |    end_key     | range_id | ...
-----------------+----------------+----------+----
  /Table/104     | /Table/104/1/3 |       56 | ...
  /Table/104/1/3 | /Table/104/2   |       57 | ...
  /Table/104/2   | /Table/104/2/3 |       55 | ...
  /Table/104/2/3 | /Table/108     |       58 | ...
  /Table/108     | /Table/109     |       59 | ...
  /Table/109     | /Max           |       60 | ...
```

New syntax: `WITH TABLES` / `WITH INDEXES`:

```
> show ranges from database defaultdb with tables; -- 1 row per range/table intersection

    start_key    |    end_key     | range_id | schema_name | table_name | table_start_key |    table_end_key     | ...
-----------------+----------------+----------+-------------+------------+-----------------+----------------------+----
  /Table/104     | /Table/104/1/3 |       56 | public      | t          | /Table/104      | /Table/105           | ...
  /Table/104/1/3 | /Table/104/2   |       57 | public      | t          | /Table/104      | /Table/105           | ...
  /Table/104/2   | /Table/104/2/3 |       55 | public      | t          | /Table/104      | /Table/105           | ...
  /Table/104/2/3 | /Table/108     |       58 | public      | t          | /Table/104      | /Table/105           | ...
  /Table/104/2/3 | /Table/108     |       58 | public      | u          | /Table/105      | /Table/106           | ...
  /Table/104/2/3 | /Table/108     |       58 | public      | v          | /Table/106      | /Table/107           | ...
  /Table/108     | /Table/109     |       59 | otherschema | w          | /Table/108      | /Table/109           | ...
  /Table/109     | /Max           |       60 | otherschema | z          | /Table/109      | /Table/109/PrefixEnd | ...
```

```
> show ranges from database defaultdb with indexes; -- 1 row per range/index intersection

    start_key    |    end_key     | range_id | schema_name | table_name | index_name | index_start_key | index_end_key | ...
-----------------+----------------+----------+-------------+------------+------------+-----------------+---------------+----
  /Table/104     | /Table/104/1/3 |       56 | public      | t          | t_pkey     | /Table/104/1    | /Table/104/2  | ...
  /Table/104/1/3 | /Table/104/2   |       57 | public      | t          | t_pkey     | /Table/104/1    | /Table/104/2  | ...
  /Table/104/2   | /Table/104/2/3 |       55 | public      | t          | sec_idx    | /Table/104/2    | /Table/104/3  | ...
  /Table/104/2/3 | /Table/108     |       58 | public      | t          | sec_idx    | /Table/104/2    | /Table/104/3  | ...
  /Table/104/2/3 | /Table/108     |       58 | public      | u          | u_pkey     | /Table/105/1    | /Table/105/2  | ...
  /Table/104/2/3 | /Table/108     |       58 | public      | v          | v_pkey     | /Table/106/1    | /Table/106/2  | ...
  /Table/108     | /Table/109     |       59 | otherschema | w          | w_pkey     | /Table/108/1    | /Table/108/2  | ...
  /Table/109     | /Max           |       60 | otherschema | z          | z_pkey     | /Table/109/1    | /Table/109/2  | ...
```

Example output for `SHOW RANGES FROM TABLE`:

```
> show ranges from table t;

   start_key   |      end_key       | range_id | ...
---------------+--------------------+----------+----
  …/<TableMin> | …/1/3              |       56 | ...
  …/1/3        | …/2                |       57 | ...
  …/2          | …/2/3              |       55 | ...
  …/2/3        | <after:/Table/108> |       58 | ...
```

```
> show ranges from table u;

         start_key        |      end_key       | range_id | ...
--------------------------+--------------------+----------+----
  <before:/Table/104/2/3> | <after:/Table/108> |       58 | ...
```

```
> show ranges from table otherschema.w;

  start_key    |  end_key     | range_id | ...
---------------+--------------+----------+----
  …/<TableMin> | …/<TableMax> |       59 | ...
```

New syntax: `SHOW RANGES FROM TABLE ... WITH INDEXES`:

```
> show ranges from table t with indexes;

  start_key    |      end_key       | range_id | index_name | index_start_key | index_end_key | ...
---------------+--------------------+----------+------------+-----------------+---------------+----
  …/<TableMin> | …/1/3              |       56 | t_pkey     | …/1             | …/2           | ...
  …/1/3        | …/<IndexMax>       |       57 | t_pkey     | …/1             | …/2           | ...
  …/<IndexMin> | …/2/3              |       55 | sec_idx    | …/2             | …/3           | ...
  …/2/3        | <after:/Table/108> |       58 | sec_idx    | …/2             | …/3           | ...
```

```
> show ranges from table u with indexes;

         start_key        |      end_key       | range_id | index_name | index_start_key | index_end_key | ...
--------------------------+--------------------+----------+------------+-----------------+---------------+----
  <before:/Table/104/2/3> | <after:/Table/108> |       58 | u_pkey     | …/1             | …/2           | ...
```

```
> show ranges from table otherschema.w with indexes;

  start_key    |  end_key     | range_id | index_name | index_start_key | index_end_key | ...
---------------+--------------+----------+------------+-----------------+---------------+----
  …/<TableMin> | …/<TableMax> |       59 | w_pkey     | …/1             | …/2           | ...
```

Example output for `SHOW RANGES FROM INDEX`:

```
> show ranges from index t@t_pkey;

  start_key    |  end_key     | range_id | ...
---------------+--------------+----------+----
  …/<TableMin> | …/3          |       56 | ...
  …/3          | …/<IndexMax> |       57 | ...
```

```
> show ranges from index t@sec_idx;

  start_key    |      end_key       | range_id | ...
---------------+--------------------+----------+----
  …/<IndexMin> | …/3                |       55 | ...
  …/3          | <after:/Table/108> |       58 | ...
```

```
> show ranges from index u@u_pkey;

         start_key        |      end_key       | range_id | ...
--------------------------+--------------------+----------+----
  <before:/Table/104/2/3> | <after:/Table/108> |       58 | ...
```

```
> show ranges from index otherschema.w@w_pkey;

  start_key    |  end_key     | range_id | ...
---------------+--------------+----------+----
  …/<TableMin> | …/<TableMax> |       59 | ...
```

See release notes below for details.

</details>

----

attention banner

**Backward-incompatible changes.**

Release note (backward-incompatible change): CockroachDB now supports
sharing storage ranges across multiple indexes/tables. As a result,
there is no more guarantee that there is at most one SQL object (e.g.
table/index/sequence/materialized view) per storage range.

Therefore, the columns `table_id`, `database_name`, `schema_name`,
`table_name` and `index_name` in `crdb_internal.ranges` and
`.ranges_no_leases` have become nonsensical: a range cannot be
attributed to a single table/index any more.

As a result:

- The aforementioned columns in the `crdb_internal` virtual tables
  have been removed. Existing code can use the SHOW RANGES
  statement instead, optionally using WITH KEYS to expose
  the raw start/end keys.

- `SHOW RANGES FROM DATABASE` continues to report one row per range,
  but stops returning the database / schema / table / index name.

- `SHOW RANGES FROM TABLE` continues to report one row per range,
  but stops returning the index name.

Suggested replacements:

- Instead of `SELECT range_id FROM crdb_internal.ranges WHERE table_name = 'x'`

  Use: `SELECT range_id FROM [SHOW RANGES FROM TABLE x]`

- Instead of `SELECT range_id FROM crdb_internal.ranges WHERE table_name = $1 OR table_id = $2`
  (variable / unpredictable table name or ID)

  Use: `SELECT range_id FROM [SHOW RANGES FROM CURRENT_CATALOG WITH TABLES] WHERE table_name = $1 OR table_id = $2`

- Instead of `SELECT start_key FROM crdb_internal.ranges WHERE table_name = 'x'`

  Use: `SELECT raw_start_key FROM [SHOW RANGES FROM TABLE x WITH KEYS]`

- Instead of `SELECT start_key FROM crdb_internal.ranges WHERE table_name = $1 OR table_id = $2`
  (unpredictable / variable table name or ID)

  Use: `SELECT raw_start_key FROM [SHOW RANGES FROM CURRENT_CATALOG WITH TABLES, KEYS] WHERE table_name = $1 OR table_id = $2`

Release note (backward-incompatible change): The format of the
columns `start_key` and `end_key` for `SHOW RANGES FROM DATABASE`
and `SHOW RANGES FROM TABLE` have been extended to include which
table/index the key belongs to. This is necessary because a range
can now contain data from more than one table/index.

Release note (backward-incompatible change): The format of
the columns `start_key` and `end_key` for `SHOW RANGE ... FOR ROW`
has been changed to stay consistent with the output of `SHOW RANGES
FROM INDEX`.

Release note (backward-incompatible change): The output of `SHOW
RANGES` does not include `range_size`, `range_size_mb`, `lease_holder`
and `lease_holder_localities` any more by default. This ensures that
`SHOW RANGES` remains fast in the common case. Use the (NEW) option
`WITH DETAILS` to include these columns.

----

**Other changes.**

Release note (bug fix): In some cases the start/end key columns of the
output of `SHOW RANGES` was missing. This was corrected.

Release note (sql change): Two new virtual tables
`crdb_internal.index_spans` and `.table_spans` have been introduced,
which list the logical keyspace used by each index/table.

----

**New features.**

Release note (sql change): The following new statements are
introduced:

- `SHOW RANGES FROM CURRENT_CATALOG` and `SHOW RANGES` without
  parameter: alias for `SHOW RANGES FROM DATABASE` on the session's
  current database.

- `SHOW RANGES FROM DATABASE ... WITH TABLES`
  Reports at least one row per table. It's possible for the same
  range ID to be repeated across multiple rows, when a range spans
  multiple tables.

- `SHOW RANGES FROM DATABASE ... WITH INDEXES`
  Reports at least one row per index. It's possible for the same
  range ID to be repeated across multiple rows, when a range spans
  multiple indexes.

- `SHOW RANGES FROM TABLE ... WITH INDEXES`
  Reports at least one row per index. It's possible for the same
  range ID to be repeated across multiple rows, when a range spans
  multiple indexes.

- `SHOW CLUSTER RANGES [ WITH { INDEXES | TABLES } ]` Reports ranges
  across the entire cluster, including ranges that don't contain table
  data. The behavior of `WITH INDEXES` and `WITH TABLES` is the same
  as for `SHOW RANGES FROM DATABASE`.

Additionally, the following new options have been added to the `SHOW
RANGES` statement:

- `WITH KEYS`: produce the raw bytes of the start/end key boundaries.
- `WITH DETAILS`: produce more details, using computations that
  require extra network roundtrips. Makes the operation slower
  overall.
- `WITH EXPLAIN`: produce the text of the SQL query used to
  run the statement.

### CODE REVIEW

**Comment by knz** [comment]
todo: update commit message with the new WITH clause

**Comment by knz** [comment]
This is good to go now.
Can I get review from:
- @ecwall for the SQL bits
- @arulajmani to review the proposed semantics (as per commit message + doc in `sql/delegate/show_ranges.go`) to see if this suitably solves the problem at hand.

cc @ajwerner since we chatted about it.

**Comment by knz** [comment]
Maybe @irfansharif you could support @arulajmani and @ecwall, since there is quite a bit to look at?

**Comment by knz** [comment]
@dt @ajwerner welcoming your advice here.
The test TestProtectedTimestampsDuringImportInto in  `sql/importer: ` fails with
```
    import_into_test.go:123: error executing '
        SELECT raw_start_key
        FROM [SHOW RANGES FROM TABLE foo WITH KEYS]
        ORDER BY raw_start_key ASC': pq: relation "foo" is offline: importing
```

That's because (presumably) SHOW RANGES uses `cat.ResolveTableIndex()` and that requires the descriptor to be PUBLIC. 
Do I have an alternative? Is there a name resolution function I can use that accepts non-public descriptors? For SHOW RANGES, that should always be OK.

(The alternative I'll use for now is `SHOW RANGES FROM DATABASE` and then filter on table name, because the filtering uses `crdb_internal.tables` which doesn't use name resolution. But I would prefer the method above instead.)

**Comment by knz** [comment]
TODO: talk to field eng team + doc team on how to teach how to use a join to get the same result from SHOW RANGES.

**Review by ecwall — COMMENTED** [review]
> Reviewed 10 of 63 files at r2, 46 of 62 files at r3, 39 of 44 files at r4, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, @miretskiy, @renatolabs, @rhu713, and @smg260)

___
*[`pkg/sql/crdb_internal.go` line 3510 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfT_aF02GzLkvZYTHP:-NJfT_aF02GzLkvZYTHQ:bhk111l) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/crdb_internal.go#L3510)):*
> ```Go
> 	comma := ""
> 	for _, c := range cols {
> 		buf.WriteString(comma)
> ```

Is this better than
```
if i > 0 {
  buf.WriteString(",")
}
```
?

It looks like the `i > 0` version is used in more places. I guess this is what happens when the language makes you implement your own string joining everywhere...
___
*[`pkg/sql/virtual_schema.go` line 219 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfcY_s7Ad3cgfw8hYP:-NJfcY_s7Ad3cgfw8hYQ:bvtlpee) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/virtual_schema.go#L219)):*
> ```Go
> 	}
> 
> 	if t.resultColumns != nil {
> ```

Can this be done as a unit test by iterating over the virtual tables or are they not all know ahead of time?
___
*[`pkg/sql/delegate/show_ranges.go` line 186 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfYZ2W1W8n7UBj2ivM:-NJfYZ2W1W8n7UBj2ivN:b-lvkciy) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/delegate/show_ranges.go#L186)):*
> ```Go
> // along with the list of replicas and the lease holder.
> // See the explanatory comment at the top of this file for details.
> func (d *delegator) delegateShowRanges(n *tree.ShowRanges) (tree.Statement, error) {
> ```

Is it worth grouping `FROM TABLE`, `FROM INDEX`, `FROM DATABASE`, etc together in this one function?

It is kind of hard to follow with so many conditionals and I think would be easier to read if there were a function per each `FROM TABLE`, `FROM INDEX`, `FROM DATABASE` that called helper functions for any shared logic.
___
*[`pkg/sql/sem/tree/show.go` line 742 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJf_plC7ri3N0-Fz9fU:-NJf_plC7ri3N0-Fz9fV:b4ey2ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/sem/tree/show.go#L742)):*
> ```Go
> const (
> 	// UniqueRanges tells to use just 1 row per range in the output.
> 	UniqueRanges ShowRangesMode = 0
> ```

`iota`?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, @miretskiy, @renatolabs, @rhu713, and @smg260)

___
*[`pkg/sql/virtual_schema.go` line 219 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfcY_s7Ad3cgfw8hYP:-NJoZNrCE2RuQOUKot81:b-87243w) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/virtual_schema.go#L219)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Can this be done as a unit test by iterating over the virtual tables or are they not all know ahead of time?
</blockquote></details>

We could do this as a unit test, but this way we get an early error even before running unit tests.
Note:
- we have precedent doing this type of check, see e.g. pkg/sql/sem/builtins/fixed_oids.go function `signatureMustHaveHardcodedOID`
- this code only runs once during server initialization, not on every query/txn.
___
*[`pkg/sql/delegate/show_ranges.go` line 186 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfYZ2W1W8n7UBj2ivM:-NJo_Pwk0Mgr_bL7n3s7:bnhhxia) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/delegate/show_ranges.go#L186)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Is it worth grouping `FROM TABLE`, `FROM INDEX`, `FROM DATABASE`, etc together in this one function?

It is kind of hard to follow with so many conditionals and I think would be easier to read if there were a function per each `FROM TABLE`, `FROM INDEX`, `FROM DATABASE` that called helper functions for any shared logic.
</blockquote></details>

That is actually what I started with, but there was a lot of redundant code between the 3 functions; and when I wanted to modify one of the stages every time I would forget to modify one of the functions. Having them together is actually easier for maintainance.

I agree it's perhaps hard to read. Added some comments to clarify. Do you like them?
___
*[`pkg/sql/sem/tree/show.go` line 742 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJf_plC7ri3N0-Fz9fU:-NJo_1axAyfkWSHaaGSz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/sem/tree/show.go#L742)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

`iota`?
</blockquote></details>

Done.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, @miretskiy, @renatolabs, @rhu713, and @smg260)

___
*[`pkg/sql/crdb_internal.go` line 3510 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfT_aF02GzLkvZYTHP:-NJoiWVuCimFbyf7XN-g:b4znplt) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/crdb_internal.go#L3510)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Is this better than
```
if i > 0 {
  buf.WriteString(",")
}
```
?

It looks like the `i > 0` version is used in more places. I guess this is what happens when the language makes you implement your own string joining everywhere...


</blockquote></details>

My version avoids an extra conditional. It's been used before too -- at least in quite a few places in the tree package.

**Review by ecwall — APPROVED** [review]
It looks good to me from the SQL side. If any issues come up as I familiarize myself with it through use then we can patch it.

> Reviewed 2 of 105 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-:-NJvULBwC2S6MtMRBmND:b-vsho6b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, @miretskiy, @renatolabs, @rhu713, and @smg260)

___
*[`pkg/sql/virtual_schema.go` line 219 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfcY_s7Ad3cgfw8hYP:-NJvHY4o8Khgi97ffyVf:b-jn9b5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/virtual_schema.go#L219)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

We could do this as a unit test, but this way we get an early error even before running unit tests.
Note:
- we have precedent doing this type of check, see e.g. pkg/sql/sem/builtins/fixed_oids.go function `signatureMustHaveHardcodedOID`
- this code only runs once during server initialization, not on every query/txn.
</blockquote></details>

Ok sounds good then.
___
*[`pkg/sql/delegate/show_ranges.go` line 186 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJfYZ2W1W8n7UBj2ivM:-NJvI4R-8QGg4VaxOqf-:b-u4xce5) ([raw file](https://github.com/cockroachdb/cockroach/blob/81f4ea90ac865f6c4cbfba8f440c67a9e6f9b6e0/pkg/sql/delegate/show_ranges.go#L186)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

That is actually what I started with, but there was a lot of redundant code between the 3 functions; and when I wanted to modify one of the stages every time I would forget to modify one of the functions. Having them together is actually easier for maintainance.

I agree it's perhaps hard to read. Added some comments to clarify. Do you like them?
</blockquote></details>

Yeah I wasn't sure if it would necessarily be better the other way, but I'm glad you looked at it both ways. 

The comments definitely help to map sections of code to parts of the SQL statement, thanks.
___
*[`pkg/sql/delegate/show_ranges.go` line 66 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJvJwVE2XWMVrIhr9dp:-NJvJwVFEUM2z7h-TDAA:b-qla1wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/bccfe40bfa00e9ba95d523d2290f908e8b995e02/pkg/sql/delegate/show_ranges.go#L66)):*
> ```Go
> //		SHOW RANGES FROM INDEX t@idx
> //		SHOW RANGES [ FROM DATABASE db | FROM CURRENT_CATALOG ]
> //	  SHOW CLUSTER RANGES
> ```

Uneven spacing.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, @miretskiy, @renatolabs, @rhu713, and @smg260)

___
*[`pkg/sql/delegate/show_ranges.go` line 66 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/93644#-NJvJwVE2XWMVrIhr9dp:-NJvnEpL3ljgO0AZ44UC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bccfe40bfa00e9ba95d523d2290f908e8b995e02/pkg/sql/delegate/show_ranges.go#L66)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Uneven spacing.
</blockquote></details>

Done.

**Comment by knz** [comment]
thank you!

bors r=ecwall

(praying for no merge skew...)
